### ISSUE TITLE
[x/programs] external program calls should return result

### ISSUE THREAD

**Issue by hexfusion** [ava-labs/hypersdk#595]
Currently exported `public` Program functions all return a i64 and most errors are panics. This issue is for returning a result and handling the custom error types.

I think it would be easier for the user to reason about something like this.
```rust
enum ProgramError {
     Fatal(String), // <- panic
     OhhNotSooBad(String), // some message that gets handled by result
}
```

possible usage

```rust
/// Initializes the program with a name, symbol, and total supply.
#[public]
pub fn init(program: Program) -> Result<(), ProgramError>{
    // set total supply
    program
        .state()
        .store(StateKey::TotalSupply.to_vec(), &123456789_i64)?;
[....]
```

**Comment by Sanghren** [comment]
As a user, I would definitely like something like this. It feels weird to have that `-> i64`. Intuitively while testing one of the example, I tried to return a `String` in one of the function and it gave me an error that had me puzzled for a while.

 `(Get name of token : PlanResponse { id: 0, result: PlanResult { id: None, msg: None, timestamp: 1698855794, response: None }, error: Some("program execution failed: invalid parameter count for function get_name: 2 expected: 3") }`.

**Comment by hexfusion** [comment]
Yeah I am working on this now. Ideally you could return whatever object you wanted as long as it was properly serializable. Stay tuned.

**Comment by richardpringle** [comment]
@hexfusion, were you thinking that `ProgramError` would be defined by the user or in the SDK?

**Comment by hexfusion** [comment]
> @hexfusion, were you thinking that `ProgramError` would be defined by the user or in the SDK?

I am still working through this but as @Sanghren noted it is supported functionality of Solidity and logically it seems that there is no reason the user can not define errors to a certain extent. Perhaps for simplicity we have concrete types exposed by the SDK that the user can chose and pass a custom `String`.

On the `HyperSDK` side of this we still need the result.Object API (not built) to adhere to some constraints as well.

**Comment by richardpringle** [comment]
> > @hexfusion, were you thinking that `ProgramError` would be defined by the user or in the SDK?
> 
> I am still working through this but as @Sanghren noted it is supported functionality of Solidity and logically it seems that there is no reason the user can not define errors to a certain extent. Perhaps for simplicity we have concrete types exposed by the SDK that the user can choose and pass a custom error `String`.
> 
> On the `HyperSDK` side of this we still need the result.Object API (not built) to adhere to some constraints as well.
> 
> What would you recommend?

I'm thinking that we can have something like this (pseudo-code)
```rust
#[thiserror::Error]
pub enum PublicError {
    #[error("state error: {0}")]
    StateError(StateError), // probably want separate get/set errors
    #[error("input parsing error")]
    Input, 
    #[error("response error: {0}")]
    Response(ResponseError), 
}

#[thiserror::Error]
pub enum ResponseError {
    SomePredefinedError,
    #[error("error produced by program: {0}")]
    Custom(Box<dyn std::error::Error>), // or something like this 
}

// with 

#[public]
fn my_fn(ctx: Ctx, input: MyInput) -> Result<Ctx, ResponseError> { .. }
```

We can have multiple layers of conversions where some conversions are hidden behind the `public` macro. Or something along those lines.

**Comment by manojkgorle** [comment]
Hey, guys. 
With reference to the issue #714 .

- Returned struct can be used to pass output values of the executed function. When a function is executed, the end user may need to know the values of some variables at the end of execution for various reasons. Viz a getter function to state variables for an Oracle price feed smart contract, the latest round details...
- we can access memory from a heap struct after freeing it at the end of the rust execution. But this can not be done for a stack struct.
We can wrap the output struct with a macro to handle all the processing.

Example implementation. 
```rust
struct current_price_feed_data {
    price: u64,
    last_round_data: u32,
}

#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

#[no_mangle]
pub extern "C" fn get_last_round_data()-> u64{
    let feed = Box::new(current_price_feed_data { x: 10, y: 20 }); //stored on the heap
    let ptr:*const current_price_feed_data= &*feed;
    let size = std::mem::size_of::<feed>();
    std::mem::forget(ptr);
    (ptr as u64) << 32 | size as u64
}
```

ig, we can also emit logs, in the same way as they gonna be some strings of formatted information.

**Comment by richardpringle** [comment]
Makes sense @manojkgorle! We can't quite do exactly what you recommended because Rust doesn't have a stable ABI, which means that the representation of that struct in binary has no guarantees between Rust versions. We could use `repr(C)`, but there's no good way to enforce that. 

Further, reading the struct on the host side could be a bit of a challenge. How is the VM supposed to know how to deserialize the bytes? 

The way things are done in the link you presented are strictly binary. I think we'll look into doing something along those lines. I'm probably going to pick this up and work on it this week.

**Comment by iFrostizz** [comment]
Now that we have https://github.com/ava-labs/hypersdk/pull/896, what stops from returning a `Result<T, wasmlanche_sdk::state::Error>` from the functions we want ? It's okay imo to let the user decide if they want to return a "system error" or not

```rust
/// Increments the count at the address by the amount.
#[public]
pub fn inc(context: Context, to: Address, amount: i64) -> Result<bool, Error> {
    let counter = amount + get_value(context, to)?;
    let Context { program, .. } = context;

    program
        .state()
        .store(StateKeys::Counter(to), &counter)
        .map_err(|_| Error::Other("failed to store counter".into()))?;

    Ok(true)
}

/// Increments the count at the address by the amount for an external program.
#[public]
pub fn inc_external(
    _: Context,
    target: Program,
    max_units: i64,
    of: Address,
    amount: i64,
) -> Result<bool, Error> {
    let params = params!(&of, &amount).unwrap();
    target.call_function("inc", &params, max_units)
}
```

The only part that is faillible from the expanded code of a `#[public]` function is the serialization of args. It would be kind of a chicken-egg problem because the `Result` would also need to be serialized, what to do in the case of this one failing too ?

**Comment by richardpringle** [comment]
>Now that we have https://github.com/ava-labs/hypersdk/pull/896, what stops us from returning a `Result<T, wasmlanche_sdk::state::Error>` from the functions we want?
 
 I'm going to assume that `T` is some placeholder for a concrete type. If that's the case, nothing is preventing us from doing that. `Result<T, E>` is borsh-(de)serializable, so we would just need to figure out the best way to represent that on the host side of things.  
 
 >It's okay imo to let the user decide if they want to return a "system error" or not
 
 _Why_ do you think that's okay? What do you gain from letting users return their own errors? Why do you think we should have a `Result` in the first place? There are many different reasons why we might want to do these things, so I'd like to hear your opinions. The answers will likely affect the implementation. 
 
 >The only part that is faillible from the expanded code of a `#[public]` function is the serialization of args. It would be kind of a chicken-egg problem because the `Result` would also need to be serialized, what to do in the case of this one failing too ?
 
 Ha! This is the exact type of thing we need to be thinking about. So, we need a response that can *always* be serialized. That likely means something like `Result<wasmlanche_sdk::Response<UserDefinedType>, wasmlanche_sdk::Error>`. Here, the user-defined type could theoretically be some enum. Conversely, we could give users an error variant in which they can store a primitive. The design space is pretty open.

**Comment by iFrostizz** [comment]
> Why do you think that's okay? What do you gain from letting users return their own errors?

It was more of a guess at a technical challenge that we may face while writing this feature. If we don't leave custom errors in the same enum as the "system errors", then it may be a pain to write failible functions because of the nesting. We could also hide this by writing some kind of static analysis within the `public` but it seems to be pretty limiting. 

I guess that we could otherwise make the `Error` enum private to the crate and write a bunch of custom functions like `unwrap_or_program_err`, `ok_or_program_err` just like `eyre` does with `ok_or_eyre` for instance and have utils like `eyre::bail!` so that the user won't be able to pick from system errors.

**Comment by richardpringle** [comment]
Relevant discussion:
https://github.com/ava-labs/hypersdk/discussions/1020

**Comment by iFrostizz** [comment]
tldr of #1020 

We need to define an error enum on the guest side like so

```rust
#[derive(thiserror::Error)]
#[non_exhaustive]
#[repr(u8)]
enum ExternalCallError {
    ExecutionFailure = 0,
    CallPanicked = 1,
}
```

The `ExecutionFailure` variant is deliberately set as all-encompassing and could be related to calling an inexistent function, an out of fuel error, and other exceptional errors due to the simulator functioning (mainly I/O with the database).
The `CallPanicked` is set to encourage program authors to fire this error if a unreachable branch has been reached or that an important invariant has been broken (not expected) but to still abort execution. It is not encouraged though to use it in any case. For instance, it is not idiomatic nor efficient to pattern match a panic message (Note that it might not even be available) so we should encourage users to return their errors through the `T` of the return type of the `call_function` which now becomes `Result<T, ExternalCallError>`.

**Comment by richardpringle** [comment]
Thanks @iFrostizz! 

I opened a new issue since the comments here aren't exactly relevant to the task: https://github.com/ava-labs/hypersdk/issues/1034

**Comment by richardpringle** [comment]
In case it's not clear, closing this in favour of https://github.com/ava-labs/hypersdk/issues/1034
